File: hu/codes/java/chapter_tree/avl_tree.java
/**
 * File: avl_tree.java
 * Created Time: 2022-12-10
 * Author: krahets (krahets@163.com)
 */

package chapter_tree;

import utils.*;

/* AVL fa */
class AVLTree {
    TreeNode root; // Gyökércsomópont

    /* Csomópont magasságának lekérése */
    public int height(TreeNode node) {
        // Üres csomópont magassága -1, levélcsomópont magassága 0
        return node == null ? -1 : node.height;
    }

    /* Csomópont magasságának frissítése */
    private void updateHeight(TreeNode node) {
        // A csomópont magassága a legmagasabb részfa magassága + 1
        node.height = Math.max(height(node.left), height(node.right)) + 1;
    }

    /* Egyensúlyi tényező lekérése */
    public int balanceFactor(TreeNode node) {
        // Üres csomópont egyensúlyi tényezője 0
        if (node == null)
            return 0;
        // Csomópont egyensúlyi tényezője = bal részfa magassága - jobb részfa magassága
        return height(node.left) - height(node.right);
    }

    /* Jobbra forgatás művelete */
    private TreeNode rightRotate(TreeNode node) {
        TreeNode child = node.left;
        TreeNode grandChild = child.right;
        // child-ot forgáspontként használva node-ot jobbra forgatja
        child.right = node;
        node.left = grandChild;
        // Csomópont magasságának frissítése
        updateHeight(node);
        updateHeight(child);
        // A forgatás utáni részfa gyökércsomópontjának visszaadása
        return child;
    }

    /* Balra forgatás művelete */
    private TreeNode leftRotate(TreeNode node) {
        TreeNode child = node.right;
        TreeNode grandChild = child.left;
        // child-ot forgáspontként használva node-ot balra forgatja
        child.left = node;
        node.right = grandChild;
        // Csomópont magasságának frissítése
        updateHeight(node);
        updateHeight(child);
        // A forgatás utáni részfa gyökércsomópontjának visszaadása
        return child;
    }

    /* Forgatás végrehajtása a részfa egyensúlyának visszaállításához */
    private TreeNode rotate(TreeNode node) {
        // A csomópont egyensúlyi tényezőjének lekérése
        int balanceFactor = balanceFactor(node);
        // Balra dőlt fa
        if (balanceFactor > 1) {
            if (balanceFactor(node.left) >= 0) {
                // Jobbra forgatás
                return rightRotate(node);
            } else {
                // Először balra, majd jobbra forgatás
                node.left = leftRotate(node.left);
                return rightRotate(node);
            }
        }
        // Jobbra dőlt fa
        if (balanceFactor < -1) {
            if (balanceFactor(node.right) <= 0) {
                // Balra forgatás
                return leftRotate(node);
            } else {
                // Először jobbra, majd balra forgatás
                node.right = rightRotate(node.right);
                return leftRotate(node);
            }
        }
        // Egyensúlyos fa, nincs szükség forgatásra, közvetlenül visszatér
        return node;
    }

    /* Csomópont beszúrása */
    public void insert(int val) {
        root = insertHelper(root, val);
    }

    /* Csomópont rekurzív beszúrása (segédmetódus) */
    private TreeNode insertHelper(TreeNode node, int val) {
        if (node == null)
            return new TreeNode(val);
        /* 1. Beszúrási pozíció keresése és csomópont beszúrása */
        if (val < node.val)
            node.left = insertHelper(node.left, val);
        else if (val > node.val)
            node.right = insertHelper(node.right, val);
        else
            return node; // Duplikált csomópont nem kerül be, közvetlenül visszatér
        updateHeight(node); // Csomópont magasságának frissítése
        /* 2. Forgatás végrehajtása a részfa egyensúlyának visszaállításához */
        node = rotate(node);
        // A részfa gyökércsomópontjának visszaadása
        return node;
    }

    /* Csomópont törlése */
    public void remove(int val) {
        root = removeHelper(root, val);
    }

    /* Csomópont rekurzív törlése (segédmetódus) */
    private TreeNode removeHelper(TreeNode node, int val) {
        if (node == null)
            return null;
        /* 1. Csomópont megkeresése és törlése */
        if (val < node.val)
            node.left = removeHelper(node.left, val);
        else if (val > node.val)
            node.right = removeHelper(node.right, val);
        else {
            if (node.left == null || node.right == null) {
                TreeNode child = node.left != null ? node.left : node.right;
                // Gyermekcsomópontok száma = 0, közvetlenül törli és visszatér
                if (child == null)
                    return null;
                // Gyermekcsomópontok száma = 1, közvetlenül törli
                else
                    node = child;
            } else {
                // Gyermekcsomópontok száma = 2, törli a következő csomópontot a középrendű bejárásban és az aktuális csomópontot azzal helyettesíti
                TreeNode temp = node.right;
                while (temp.left != null) {
                    temp = temp.left;
                }
                node.right = removeHelper(node.right, temp.val);
                node.val = temp.val;
            }
        }
        updateHeight(node); // Csomópont magasságának frissítése
        /* 2. Forgatás végrehajtása a részfa egyensúlyának visszaállításához */
        node = rotate(node);
        // A részfa gyökércsomópontjának visszaadása
        return node;
    }

    /* Csomópont keresése */
    public TreeNode search(int val) {
        TreeNode cur = root;
        // Ciklikus keresés, levélcsomóponton túllépve kilép
        while (cur != null) {
            // A célcsomópont cur jobb részfájában van
            if (cur.val < val)
                cur = cur.right;
            // A célcsomópont cur bal részfájában van
            else if (cur.val > val)
                cur = cur.left;
            // Megtalálta a célcsomópontot, kilép a ciklusból
            else
                break;
        }
        // A célcsomópont visszaadása
        return cur;
    }
}

public class avl_tree {
    static void testInsert(AVLTree tree, int val) {
        tree.insert(val);
        System.out.println("\nAfter inserting node " + val + ", AVL tree is");
        PrintUtil.printTree(tree.root);
    }

    static void testRemove(AVLTree tree, int val) {
        tree.remove(val);
        System.out.println("\nAfter removing node " + val + ", AVL tree is");
        PrintUtil.printTree(tree.root);
    }

    public static void main(String[] args) {
        /* Kérjük, figyelje meg, hogyan tartja fenn az AVL fa az egyensúlyt csomópontok beszúrása után */
        AVLTree avlTree = new AVLTree();

        /* Csomópont beszúrása */
        // Csomópontok törlése
        testInsert(avlTree, 1);
        testInsert(avlTree, 2);
        testInsert(avlTree, 3);
        testInsert(avlTree, 4);
        testInsert(avlTree, 5);
        testInsert(avlTree, 8);
        testInsert(avlTree, 7);
        testInsert(avlTree, 9);
        testInsert(avlTree, 10);
        testInsert(avlTree, 6);

        /* Kérjük, figyelje meg, hogyan tartja fenn az AVL fa az egyensúlyt csomópontok törlése után */
        testInsert(avlTree, 7);

        /* Csomópont törlése */
        // 1 fokú csomópont törlése
        testRemove(avlTree, 8); // 2 fokú csomópont törlése
        testRemove(avlTree, 5); // 1 fokú csomópont törlése
        testRemove(avlTree, 4); // 2 fokú csomópont törlése

        /* Csomópont keresése */
        TreeNode node = avlTree.search(7);
        System.out.println("\nFound node object is " + node + ", node value = " + node.val);
    }
}

File: hu/codes/python/chapter_tree/avl_tree.py
"""
File: avl_tree.py
Created Time: 2022-12-20
Author: a16su (lpluls001@gmail.com)
"""

import sys
from pathlib import Path

sys.path.append(str(Path(__file__).parent.parent))
from modules import TreeNode, print_tree


class AVLTree:
    """AVL fa"""

    def __init__(self):
        """Konstruktor"""
        self._root = None

    def get_root(self) -> TreeNode | None:
        """A bináris fa gyökércsomópontjának lekérése"""
        return self._root

    def height(self, node: TreeNode | None) -> int:
        """Csomópont magasságának lekérése"""
        # Az üres csomópont magassága -1, a levélcsomóponté 0
        if node is not None:
            return node.height
        return -1

    def update_height(self, node: TreeNode | None):
        """Csomópont magasságának frissítése"""
        # A csomópont magassága = a legmagasabb részfa magassága + 1
        node.height = max([self.height(node.left), self.height(node.right)]) + 1

    def balance_factor(self, node: TreeNode | None) -> int:
        """Egyensúlyi tényező lekérése"""
        # Az üres csomópont egyensúlyi tényezője 0
        if node is None:
            return 0
        # Csomópont egyensúlyi tényezője = bal részfa magassága - jobb részfa magassága
        return self.height(node.left) - self.height(node.right)

    def right_rotate(self, node: TreeNode | None) -> TreeNode | None:
        """Jobbra forgatás"""
        child = node.left
        grand_child = child.right
        # A child-ot forgáspontként használva a node-ot jobbra forgatja
        child.right = node
        node.left = grand_child
        # Csomópont magasságának frissítése
        self.update_height(node)
        self.update_height(child)
        # A forgatás utáni részfa gyökércsomópontját adja vissza
        return child

    def left_rotate(self, node: TreeNode | None) -> TreeNode | None:
        """Balra forgatás"""
        child = node.right
        grand_child = child.left
        # A child-ot forgáspontként használva a node-ot balra forgatja
        child.left = node
        node.right = grand_child
        # Csomópont magasságának frissítése
        self.update_height(node)
        self.update_height(child)
        # A forgatás utáni részfa gyökércsomópontját adja vissza
        return child

    def rotate(self, node: TreeNode | None) -> TreeNode | None:
        """Forgatás végrehajtása a részfa egyensúlyának helyreállításához"""
        # Csomópont egyensúlyi tényezőjének lekérése
        balance_factor = self.balance_factor(node)
        # Bal oldali túlsúly
        if balance_factor > 1:
            if self.balance_factor(node.left) >= 0:
                # Jobbra forgatás
                return self.right_rotate(node)
            else:
                # Előbb balra, majd jobbra forgatás
                node.left = self.left_rotate(node.left)
                return self.right_rotate(node)
        # Jobb oldali túlsúly
        elif balance_factor < -1:
            if self.balance_factor(node.right) <= 0:
                # Balra forgatás
                return self.left_rotate(node)
            else:
                # Előbb jobbra, majd balra forgatás
                node.right = self.right_rotate(node.right)
                return self.left_rotate(node)
        # Kiegyensúlyozott fa, nincs szükség forgatásra, közvetlenül visszaadja
        return node

    def insert(self, val):
        """Csomópont beszúrása"""
        self._root = self.insert_helper(self._root, val)

    def insert_helper(self, node: TreeNode | None, val: int) -> TreeNode:
        """Rekurzív csomópont-beszúrás (segédmetódus)"""
        if node is None:
            return TreeNode(val)
        # 1. Beszúrási pozíció keresése és csomópont beszúrása
        if val < node.val:
            node.left = self.insert_helper(node.left, val)
        elif val > node.val:
            node.right = self.insert_helper(node.right, val)
        else:
            # Duplikált csomópont nem kerül be, közvetlenül visszaadja
            return node
        # Csomópont magasságának frissítése
        self.update_height(node)
        # 2. Forgatás végrehajtása a részfa egyensúlyának helyreállításához
        return self.rotate(node)

    def remove(self, val: int):
        """Csomópont törlése"""
        self._root = self.remove_helper(self._root, val)

    def remove_helper(self, node: TreeNode | None, val: int) -> TreeNode | None:
        """Rekurzív csomópont-törlés (segédmetódus)"""
        if node is None:
            return None
        # 1. Csomópont keresése és törlése
        if val < node.val:
            node.left = self.remove_helper(node.left, val)
        elif val > node.val:
            node.right = self.remove_helper(node.right, val)
        else:
            if node.left is None or node.right is None:
                child = node.left or node.right
                # Gyermekek száma = 0, a csomópontot közvetlenül törli és None-t ad vissza
                if child is None:
                    return None
                # Gyermekek száma = 1, a csomópontot közvetlenül törli
                else:
                    node = child
            else:
                # Gyermekek száma = 2, a közrendű bejárásban következő csomópontot törli, majd az aktuálist azzal helyettesíti
                temp = node.right
                while temp.left is not None:
                    temp = temp.left
                node.right = self.remove_helper(node.right, temp.val)
                node.val = temp.val
        # Csomópont magasságának frissítése
        self.update_height(node)
        # 2. Forgatás végrehajtása a részfa egyensúlyának helyreállításához
        return self.rotate(node)

    def search(self, val: int) -> TreeNode | None:
        """Csomópont keresése"""
        cur = self._root
        # Ciklusos keresés, a levélcsomópont után kilép
        while cur is not None:
            # A célcsomópont a cur jobb részfájában van
            if cur.val < val:
                cur = cur.right
            # A célcsomópont a cur bal részfájában van
            elif cur.val > val:
                cur = cur.left
            # Megtalálta a célcsomópontot, kilép a ciklusból
            else:
                break
        # A célcsomópontot adja vissza
        return cur


"""Fő kód"""
if __name__ == "__main__":

    def test_insert(tree: AVLTree, val: int):
        tree.insert(val)
        print("\A(z) {} csomópont beszúrása után az AVL fa:".format(val))
        print_tree(tree.get_root())

    def test_remove(tree: AVLTree, val: int):
        tree.remove(val)
        print("\nA(z) {} csomópont törlése után az AVL fa:".format(val))
        print_tree(tree.get_root())

    # Üres AVL fa inicializálása
    avl_tree = AVLTree()

    # Csomópontok beszúrása
    # Figyelje meg, hogyan tartja egyensúlyban az AVL fa magát csomópontok beszúrása után
    for val in [1, 2, 3, 4, 5, 8, 7, 9, 10, 6]:
        test_insert(avl_tree, val)

    # Duplikált csomópont beszúrása
    test_insert(avl_tree, 7)

    # Csomópontok törlése
    # Figyelje meg, hogyan tartja egyensúlyban az AVL fa magát csomópontok törlése után
    test_remove(avl_tree, 8)  # 0 fokú csomópont törlése
    test_remove(avl_tree, 5)  # 1 fokú csomópont törlése
    test_remove(avl_tree, 4)  # 2 fokú csomópont törlése

    result_node = avl_tree.search(7)
    print("\nA talált csomópont objektum: {}, csomópont értéke = {}".format(result_node, result_node.val))
